torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -630,6 +636,22 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   void logWorkEnd(WorkNCCL& work);
 
  protected:
+  // Function that runs as part of a separate thread aside from watchdog
+  // thread because we need to check the heartbeat from watchdog thread
+  // so that when we get stuck in some NCCL/CUDA calls,
+  // we can dump the debugging information and abort the process.
+  virtual void heartbeatMonitor();
+
+  // Function that directly trigger std::abort so that the whole process
+  // gets terminated.
+  virtual void terminateProcess(std::string errMsg);
```

Comment:
nit: should these two functions be more like private than protected?

Reply:
We cannot override a private method... I agree with you but this is c++, so we cannot do things like in Python.